From 64142908a1eb9bc70c07dcf56079cf37ebf7cebf Mon Sep 17 00:00:00 2001
From: Jason Liu <jason.hui.liu@nxp.com>
Date: Fri, 27 Jul 2018 15:43:08 +0800
Subject: [PATCH 4301/5242] MLK-19078 dma: fsl-edma-v3: set the swap correctly

commit  1e78cec700650fc255e7a8692928282b68f03890 from
https://source.codeaurora.org/external/imx/linux-imx.git

The swap is swapped between the i.MX8QM RevA and RevB
this patch handle this difference to set swap correctly
otherwise, the eDMA will not work on the i.MX8QM RevB

Signed-off-by: Jason Liu <jason.hui.liu@nxp.com>
Reviewed-by: Anson Huang <anson.huang@nxp.com>
(cherry picked from commit 5f4d3549e5f61cb8e3c14dbeb406acfcccf32886)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/fsl-edma-v3.c |   17 ++++++++++++++++-
 1 file changed, 16 insertions(+), 1 deletion(-)

diff --git a/drivers/dma/fsl-edma-v3.c b/drivers/dma/fsl-edma-v3.c
index a7b57f4..f74f46a 100644
--- a/drivers/dma/fsl-edma-v3.c
+++ b/drivers/dma/fsl-edma-v3.c
@@ -28,6 +28,9 @@
 #include <linux/of_irq.h>
 #include <linux/of_dma.h>
 
+#include <soc/imx/revision.h>
+#include <soc/imx8/soc.h>
+
 #include "virt-dma.h"
 
 #define EDMA_CH_CSR			0x00
@@ -844,9 +847,21 @@ static int fsl_edma3_probe(struct platform_device *pdev)
 	if (of_property_read_bool(np, "shared-interrupt"))
 		irqflag = IRQF_SHARED;
 
-	fsl_edma3->swap = of_device_is_compatible(np, "fsl,imx8qm-adma");
+	fsl_edma3->swap = false;
 	fsl_edma3->n_chans = chans;
 
+	/*
+	 * FIXUP: if this is the i.MX8QM TO1.0, need set the swap.
+	 * FIXME: This will be revisted to set the swap property
+	 * from the device-tree node later instead of revison check,
+	 * but, this will need add extra DT file, not perfect too.
+	 */
+
+	if ((of_device_is_compatible(np, "fsl,imx8qm-adma")) &&
+		cpu_is_imx8qm() &&
+		imx8_get_soc_revision() == IMX_CHIP_REVISION_1_0)
+		fsl_edma3->swap = true;
+
 	INIT_LIST_HEAD(&fsl_edma3->dma_dev.channels);
 	for (i = 0; i < fsl_edma3->n_chans; i++) {
 		struct fsl_edma3_chan *fsl_chan = &fsl_edma3->chans[i];
-- 
1.7.9.5

